### PR TITLE
UserStory #25 - Display reviews

### LINKED ISSUE #145: SP: 3 Create serializer for reviews GET method
Part of user story #25
Create additional serializer for getting all and by id.

### LINKED ISSUE #147: SP: 3 Create tests for display review
Part of user story #25

### LINKED ISSUE #169: SP: 3 Create GET method view for reviews
Part of user story #25

### PR DESCRIPTION
**Pull request by OleksiiDatsiuk**
I have implemented: 

- Serializer for displaying info about review
- View with GET method provided to get reviews for specified specialist
- Integration with flake8

TODO:

- Write tests for these methods
- Resolve conflicts

Closes #145 
Closes #169 
Closes #147

### CODE REVIEW

File: beauty/api/serializers/review_serializers.py [hunk 1]
```diff
@@ -1,13 +1,43 @@
...
 
 
+class ReviewDisplaySerializer(serializers.ModelSerializer):
+    """This is a serializer for review display."""
+
+    class Meta:
+        """This is a class Meta that keeps settings for serializer."""
+
+        model = Review
+        fields = ("text_body",
```
Comment (morento101) [inline beauty/api/serializers/review_serializers.py]: fields = "__all__""

File: beauty/api/serializers/review_serializers.py [hunk 2]
```diff
@@ -1,13 +1,43 @@
...
+                  "rating",
+                  "from_user",
+                  "to_user",
+                  "date_of_publication")
+
+
+class ReviewDisplayDetailSerializer(ReviewDisplaySerializer):
+    """This serializer receives data from PUT, PATCH methods and checks whether data is valid."""
+
+    def update(self, instance: object, validated_data: dict) -> object:
```
Comment (morento101) [inline beauty/api/serializers/review_serializers.py]: No need to overwrite the update method if you don't add any additional functionality

File: beauty/api/urls.py [hunk 3]
```diff
@@ -1,25 +1,39 @@
...
 urlpatterns = [
-    path('users/', CustomUserListCreateView.as_view(),
-         name='user-list-create'),
-    path('user/<int:pk>/', CustomUserDetailRUDView.as_view(),
-         name='user-detail'),
-    path('user/<int:user>/order/<int:pk>/',
+    path("", CustomUserListCreateView.as_view(), name="user-list"),
+    path("<int:pk>/", CustomUserDetailRUDView.as_view(), name="user-detail"),
+    # path('<int:user>/order/<int:id>/',
+    #      CustomUserOrderDetailRUDView.as_view(),
```
Comment (morento101) [inline beauty/api/urls.py]: delete this piece of code, if it is useless

File: beauty/api/views.py [hunk 4]
```diff
@@ -1,27 +1,31 @@
...
 
+
 from django.db.models import Q
 from django.shortcuts import redirect
 from django.utils.encoding import force_str
 from django.utils.http import urlsafe_base64_decode
 from rest_framework import status
-from rest_framework.generics import (ListCreateAPIView, get_object_or_404,
-                                     GenericAPIView,
-                                     RetrieveUpdateDestroyAPIView)
```
Comment (morento101) [inline beauty/api/views.py]: I don't understand the reason for deleting this

File: beauty/api/views.py [hunk 5]
```diff
@@ -1,27 +1,31 @@
...
 from rest_framework.permissions import (IsAuthenticatedOrReadOnly,
                                         IsAuthenticated)
-from rest_framework.permissions import IsAuthenticated
 
+from .models import CustomUser, Order, Review
+from .permissions import IsAccountOwnerOrReadOnly
+from .serializers.customuser_serializers import CustomUserDetailSerializer
+from .serializers.customuser_serializers import CustomUserSerializer
+# from .serializers.customuser_serializers import UserOrderDetailSerializer
+from .serializers.customuser_serializers import ResetPasswordSerializer
```
Comment (morento101) [inline beauty/api/views.py]: Uncomment or delete

File: beauty/beauty/urls.py [hunk 6]
```diff
@@ -50,7 +50,7 @@ def api_root(request, format=None):
     # path('api/v1/auth/', include('djoser.urls')),
     # path('api/v1/auth_token/', include('djoser.urls.authtoken')),
     path('api/v1/', include('api.urls', namespace="api")),
-    # path(
+    #path(
```
Comment (morento101) [inline beauty/beauty/urls.py]: Uncomment or delete

File: beauty/api/views.py (hunk 5 above)
Comment (morento101) [inline beauty/api/views.py]: WET - we enjoy typing

4 same imports in a row :sunglasses:
```py
from .serializers.customuser_serializers import CustomUserDetailSerializer
from .serializers.customuser_serializers import CustomUserSerializer
# from .serializers.customuser_serializers import UserOrderDetailSerializer
from .serializers.customuser_serializers import ResetPasswordSerializer
```

**Review by morento101 — CHANGES_REQUESTED** [review]
In general, good, but some not critical changes needed

File: beauty/beauty/settings.py [hunk 7]
```diff
@@ -207,6 +207,9 @@
     "DEFAULT_AUTHENTICATION_CLASSES": (
         "rest_framework_simplejwt.authentication.JWTAuthentication",
     ),
+    'DEFAULT_PAGINATION_CLASS':
+        'rest_framework.pagination.LimitOffsetPagination',
```
Comment (pazuzu-ua) [inline beauty/beauty/settings.py]: flake8: please, add double quotes

File: beauty/api/serializers/review_serializers.py [hunk 8]
```diff
@@ -1,13 +1,43 @@
...
+
+        model = Review
+        fields = ("text_body",
+                  "rating",
+                  "from_user",
+                  "to_user",
+                  "date_of_publication")
+
+
+class ReviewDisplayDetailSerializer(ReviewDisplaySerializer):
```
Comment (pazuzu-ua) [inline beauty/api/serializers/review_serializers.py]: Will you add any functionality here? Is there a need in a distinct class?

File: beauty/api/views.py [hunk 9]
```diff
@@ -230,28 +231,51 @@ def send_signal(self, order: object, request: dict) -> None:
...
+    filter_backends = (filters.OrderingFilter, )
+    ordering_fields = ("date_of_publication", )
+    ordering = ("-date_of_publication", )
+
+    def get(self, request, pk):
+        """Method for retrieving reviews from the database."""
+        queryset = self.queryset.filter(to_user=pk)
+        queryset = super().filter_queryset(queryset)
+        if not queryset:
+            logger.info(f"Reviews for user with id {pk} were unsuccessfully obtained")
```
Comment (pazuzu-ua) [inline beauty/api/views.py]: "were unsuccessfully obtained" means that you got the reviews, but you are not happy with that. Change to "failed to get reviews for user (...)"

**Review by pazuzu-ua — CHANGES_REQUESTED** [review]

File: beauty/api/urls.py [hunk 10]
```diff
@@ -1,25 +1,39 @@
...
-         name='order-approving'),
-    
-    path(r'<int:user>/reviews/add/', ReviewAddView.as_view(),
-         name='review-add'),
+         name="order-approving"),
+
+    path(r"<int:user>/reviews/add/", ReviewAddView.as_view(),
+         name="review-add"),
+    path("reviews/<int:pk>", ReviewDisplayView.as_view(),
+         name="review-get"),
```
Comment (therobotisnotatoy) [inline beauty/api/urls.py]: I think this file is not up to date. Did you marge dev into your branch?

File: beauty/api/views.py (hunk 4 above)
Comment (therobotisnotatoy) [reply beauty/api/views.py]: I think this is part of old file version. Please be updated with develop branch

**Review by therobotisnotatoy — CHANGES_REQUESTED** [review]

File: beauty/api/serializers/review_serializers.py (hunk 1 above)
Comment (OleksiiDatsiuk) [reply beauty/api/serializers/review_serializers.py]: Got it

File: beauty/api/views.py (hunk 4 above)
Comment (OleksiiDatsiuk) [reply beauty/api/views.py]: It is a part of older version of the project, fixed it by merging develop branch into mine

File: beauty/api/views.py (hunk 5 above)
Comment (OleksiiDatsiuk) [reply beauty/api/views.py]: Outdated

File: beauty/beauty/urls.py (hunk 6 above)
Comment (OleksiiDatsiuk) [reply beauty/beauty/urls.py]: Outdated

File: beauty/api/views.py (hunk 5 above)
Comment (OleksiiDatsiuk) [reply beauty/api/views.py]: Also was not my idea, I've got this, when merged develop branch into mine. The latest version doesn't include this code

File: beauty/beauty/settings.py (hunk 7 above)
Comment (OleksiiDatsiuk) [reply beauty/beauty/settings.py]: Got it

File: beauty/api/serializers/review_serializers.py (hunk 8 above)
Comment (OleksiiDatsiuk) [reply beauty/api/serializers/review_serializers.py]: For now - I am going to remove it in the next commit, cuz it is useless

File: beauty/api/views.py (hunk 9 above)
Comment (OleksiiDatsiuk) [reply beauty/api/views.py]: Got it

File: beauty/api/urls.py (hunk 10 above)
Comment (OleksiiDatsiuk) [reply beauty/api/urls.py]: Yep, all of these files were not up to date, will fix it soon

File: .gitignore [hunk 11]
```diff
@@ -3,8 +3,6 @@ __pycache__/
 *.py[cod]
 *$py.class
 .idea/
-.vscode/*
-.vscode
 .flake8
```
Comment (morento101) [inline .gitignore]: I have added those lines for ignoring my vscode project settings folder. I don't want to push them into our project repo.

Maybe, there is a reason for that?

**Review by morento101 — COMMENTED** [review]
Pls, don't kill .gitkeep :innocent:

Fix pagination, add vscode in gitignore and I will approve this pull request

**Review by datsiuk-ira — APPROVED** [review]

**Review by therobotisnotatoy — APPROVED** [review]

File: beauty/api/serializers/review_serializers.py [hunk 12]
```diff
@@ -1,40 +1,40 @@
...
+        """This is a class Meta that keeps settings for serializer."""
+
+        model = Review
+        fields = "__all__"
+
+
 class ReviewAddSerializer(serializers.ModelSerializer):
     """This is a serializer for creating a Review."""
 
+    text_body = serializers.CharField(max_length=500)
```
Comment (pazuzu-ua) [inline beauty/api/serializers/review_serializers.py:23]: There is no need in them, as it appeared to be, so you can leave them =)

File: beauty/api/serializers/review_serializers.py [hunk 13]
```diff
@@ -1,40 +1,40 @@
...
-        reviewee = kwargs["to_user"]
-        if reviewer == reviewee:
-            logger.info(f"User {reviewer} (id = {reviewer.id}) tried reviewing himself or herself.")
-            raise serializers.ValidationError(
-                {"error": "You are not able to review yourself."},
-            )
-        if not reviewee.is_specialist:
-            logger.info(f"User {reviewer} (id = {reviewer.id}) tried reviewing a non "
-                        f"specialist user {reviewee} (id = {reviewee.id}).")
+        """The save method to check whether users are trying to review themselves."""
```
Comment (pazuzu-ua) [inline beauty/api/serializers/review_serializers.py:33]: Please, check that I have changed this method, don't override with an old version

File: beauty/api/tests/factories.py [hunk 14]
```diff
@@ -86,7 +86,7 @@ class Meta:
     name = factory.Sequence(lambda n: f"Position_{n}")
     business = factory.SubFactory(BusinessFactory)
     start_time = factory.LazyFunction(timezone.now)
-    end_time = factory.LazyAttribute(lambda o: o.start_time + timezone.timedelta(hours=4))
+    end_time = factory.LazyAttribute(lambda o: o.start_time - timezone.timedelta(hours=8))
```
Comment (pazuzu-ua) [inline beauty/api/tests/factories.py:89]: Why it has "-"? End time should be later

File: beauty/api/tests/test_GET_method_to_get_reviews.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+class TestReviewDisplayView(TestCase):
+    """This class represents a Test case and has all the tests."""
+
+    def setUp(self) -> None:
+        """This method adds needed info for tests."""
+        self.user = CustomUserFactory.create(is_active=True)
+        self.first_responder = CustomUserFactory(first_name="UserSpecialist")
+        self.second_responder = CustomUserFactory(first_name="UserSpecialist2")
+
+        self.review = ReviewFactory.create(from_user=self.user, to_user=self.first_responder)
```
Comment (pazuzu-ua) [inline beauty/api/tests/test_GET_method_to_get_reviews.py]: Why you do it 4 times in a row and assign to the same variable? If you need some of them to be in a db, just do a batch_create

File: beauty/api/tests/test_GET_method_to_get_reviews.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+from .factories import CustomUserFactory, ReviewFactory
+from api.serializers.review_serializers import ReviewDisplaySerializer
+
+
+class TestReviewDisplayView(TestCase):
+    """This class represents a Test case and has all the tests."""
+
+    def setUp(self) -> None:
+        """This method adds needed info for tests."""
+        self.user = CustomUserFactory.create(is_active=True)
```
Comment (pazuzu-ua) [inline beauty/api/tests/test_GET_method_to_get_reviews.py]: There is no more need in is_active, I have fixed the factory

**Comment by OleksiiDatsiuk** [comment]
I have completely broken this pull request :) Will fix it soon

**Pull request closed without merging**